projects/extension/ai/semantic_catalog.py

```diff
@@ -0,0 +1,433 @@
+    """
+    messages = [{"role": "user", "content": message_content}]
+    tools = [
+        {
+            "name": "generate_description",
+            "description": "Record description of the columns of the table or view in well-formed JSON",
+            "input_schema": {
+                "type": "object",
+                "properties": {
+                    "columns": {
+                        "type": "array",
+                        "items": {
+                            "type": "object",
+                            "properties": {
+                                "name": {
+                                    "type": "string",
+                                    "description": "Name of the column",
+                                },
+                                "description": {
+                                    "type": "string",
+                                    "description": "Description of the column",
+                                },
+                            },
+                        },
+                    }
+                },
+            },
+        }
+    ]
+    provider = parsed_config.get("provider", None)
+    if provider == "anthropic":
+        model = parsed_config.get("model", "claude-3-5-sonnet-latest")
+        plan = plpy.prepare(
+            """
+                SELECT ai.anthropic_generate(
+                    $1
+                    , $2
+                    , system_prompt => $3
+                    , tools => $4
+                    , tool_choice => '{{"type": "tool", "name": "generate_description"}}'::jsonb
+                )
+            """,
+            ["text", "jsonb", "text", "jsonb"],
+        )
+        result = plpy.execute(
+            plan, [model, json.dumps(messages), system_prompt, json.dumps(tools)]
+        )
+        columns = json.loads(result[0]["anthropic_generate"])["content"][0]["input"][
+            "columns"
+        ]
+
+    elif provider == "ollama":
+        pass
+    elif provider == "openai":
+        model = parsed_config.get("model", "gpt-4o")
+        messages = [
+            {"role": "system", "content": system_prompt},
+            {"role": "user", "content": message_content},
+        ]
+        plan = plpy.prepare(
+            """
+                select ai.openai_chat_complete(
+                    $1
+                    , $2
+                    , tools => $3
+                    , tool_choice => '{{"type": "function", "function": {{"name": "generate_description"}}}}'
+                )
+            """,
+            ["text", "jsonb", "jsonb"],
+        )
+        result = plpy.execute(
+            plan, [model, json.dumps(messages), json.dumps(map_tools_to_openai(tools))]
+        )
+        response = json.loads(result[0]["openai_chat_complete"])
+        columns = json.loads(
+            response["choices"][0]["message"]["tool_calls"][0]["function"]["arguments"]
+        )["columns"]
+    elif provider == "cohere":
+        pass
```

Comment:
raise exception

Reply:
Done in 27853696d476c88c7c8ae47fa003081e711dc8ed